src/picongpu/include/plugins/ChargeConservation.tpp

```diff
@@ -0,0 +1,228 @@
+/**
+ * Copyright 2013-2015 Axel Huebl, Heiko Burau, Rene Widera, Felix Schmitt
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "math/vector/Int.hpp"
+#include "math/vector/Float.hpp"
+#include "math/vector/Size_t.hpp"
+#include "math/vector/math_functor/abs.hpp"
+#include "math/vector/math_functor/max.hpp"
+#include "cuSTL/container/PseudoBuffer.hpp"
+#include "dataManagement/DataConnector.hpp"
+#include "fields/FieldJ.hpp"
+#include "math/Vector.hpp"
+#include "cuSTL/algorithm/mpi/Gather.hpp"
+#include "cuSTL/algorithm/mpi/Reduce.hpp"
+#include "cuSTL/container/DeviceBuffer.hpp"
+#include "cuSTL/container/HostBuffer.hpp"
+#include "cuSTL/algorithm/kernel/Foreach.hpp"
+#include "cuSTL/algorithm/host/Foreach.hpp"
+#include "cuSTL/cursor/NestedCursor.hpp"
+#include "lambda/Expression.hpp"
+#include <sstream>
+#include "algorithms/ForEach.hpp"
+#include "cuSTL/algorithm/kernel/Reduce.hpp"
+#include "nvidia/functors/Add.hpp"
+#include "common/txtFileHandling.hpp"
+
+namespace picongpu
+{
+
+ChargeConservation::ChargeConservation()
+    : name("ChargeConservation: Print the maximum charge deviation between particles and div E to textfile 'chargeConservation.dat'"),
+      prefix("chargeConservation"), filename("chargeConservation.dat")
+{
+    Environment<>::get().PluginConnector().registerPlugin(this);
+}
+
+void ChargeConservation::pluginRegisterHelp(po::options_description& desc)
+{
+    desc.add_options()
+        ((this->prefix + ".period").c_str(),
+        po::value<uint32_t > (&this->notifyPeriod)->default_value(0), "enable plugin [for each n-th step]");
+}
+
+std::string ChargeConservation::pluginGetName() const {return this->name;}
+
+void ChargeConservation::pluginLoad()
+{
+    Environment<>::get().PluginConnector().setNotificationPeriod(this, this->notifyPeriod);
+}
+
+void ChargeConservation::restart(uint32_t restartStep, const std::string restartDirectory)
+{
+    std::ofstream output_file(this->filename.c_str());
+    restoreTxtFile( output_file,
+                    this->filename,
+                    restartStep,
+                    restartDirectory );
+}
+
+void ChargeConservation::checkpoint(uint32_t currentStep, const std::string checkpointDirectory)
+{
+    std::ofstream output_file(this->filename.c_str());
+    checkpointTxtFile( output_file,
```

Comment:
This is still performed by **all mpi ranks**. Please use only one rank as e.g. used [here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f31b31ab5afe8951ba58754fc6266a4ac07ffd20/src/picongpu/include/plugins/BinEnergyParticles.hpp#L367-L387).